Implement checkout wrapper with parameter for partial checkout

wyrnat opened the pull request:
[why]
This implementation serves as a preparation to exchange sequences between VXFEL and XFEL repository.

[what]
A git_checkout_options instance is created and the path specifier is implemented within. I used C type conversion for std::string to git_strarray, probably there is a better way to do this.
To receive the latest commit of said branch, the function checkout assumes that within repo_ variable, there is already the branch we want to checkout from. If not, the Error handling is done already in the function get_commit.
After we retrieved the branch, we collect the whole tree to be able to use the libgit2 function git_checkout_tree.
For this, it was necessary to create a wrapper for git_commit_tree

[next]
I will also implement tests for this. I decided to already open a pull request so that you can  "add your mustard" as soon as possible.
And documentation of the code...

wyrnat committed:
raw checkout which modifies the repo_ member variable inside Repository object

Finii commented on src/Repository.cc:
```diff
@@ -594,6 +594,27 @@ bool Repository::branch_up_to_date(const std::string& branch_name)
 }
 #endif
 
+void Repository::checkout(const std::string& branch_name, const std::vector<std::string>& paths)
```
Signature different from header file where it is 
```c++
    /**
     * Checkout a branch with options
    */
    void checkout(const std::string& branch_name, const std::string& paths = "*");
```
(vector versus signle string)

Finii replied:
I changed this for my tests to

```diff
diff --git a/include/libgit4cpp/Repository.h b/include/libgit4cpp/Repository.h
index c0f7faf..5b36f0f 100644
--- a/include/libgit4cpp/Repository.h
+++ b/include/libgit4cpp/Repository.h
@@ -239,7 +239,7 @@ public:
     /**
      * Checkout a branch with options
     */
-    void checkout(const std::string& branch_name, const std::string& paths = "*");
+    void checkout(const std::string& branch_name, const std::vector<std::string>& paths = {"*"});
 
     /**
      * Remove all entries from the index under a given directory.
```

Finii commented on src/Repository.cc:
```diff
@@ -594,6 +594,27 @@ bool Repository::branch_up_to_date(const std::string& branch_name)
+    git_checkout_options checkout_opts = GIT_CHECKOUT_OPTIONS_INIT;
+    // define the paths of files to checkout
+    checkout_opts.paths.count = paths.size();
+    checkout_opts.paths.strings = (char **) paths.data();
```
C style cast

wyrnat replied:
A review was not planned yet, I was just requesting a merge so that Lars can see if that is what he needs. Probably I should have asked him eye-to-eye. But thanks anyway, I will prepare for review tomorrow.

alt-graph replied:
> A review was not planned yet, I was just requesting a merge so that Lars can see if that is what he needs. Probably I should have asked him eye-to-eye. But thanks anyway, I will prepare for review tomorrow.

@wyrnat: I have converted this into a draft as long as it is not ready for review. Please mark it as "ready for review" whenever you are ready.

Finii replied:
> I will prepare for review tomorrow.

Please define `tomorrow` ;-D

Just joking.

wyrnat replied:
@Finii tomorrow is today :P

Finii reviewed (commented):
This seems to be not ready for review?

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/5c4a315a-91c8-4a05-b46a-8c9e8b2b2a78)

wyrnat committed:
add new_branch and reference_shorthand

wyrnat committed:
add git_branch_create wrapper

wyrnat committed:
debug finished. TODO: tests

wyrnat committed:
transform specification into reference and reference into longname

wyrnat committed:
write test to switch branch and add switch_branch function to Repository

wyrnat committed:
finish partial checkout test

alt-graph commented on .gitignore:
```diff
@@ -4,3 +4,4 @@ docs/html/
 .vscode/settings.json
 .vscode/cspell.json
 unit_test_files/
+sequences/
```
All unit test stuff should be generated under "unit_test_files/", so why exclude "sequences/"?

wyrnat replied:
I still had this folder in my project, I will delete it

alt-graph commented on .vscode/c_cpp_properties.json:
```diff
@@ -4,6 +4,7 @@
             "name": "Ubuntu-x64",
             "includePath": [
                 "${workspaceFolder}/**",
+                "/export/doocs/lib/include",
```
This line is duplicated?

wyrnat replied:
Yes, and I dont know how it got there. I did not add it manually.

alt-graph commented on include/libgit4cpp/Repository.h:
```diff
@@ -233,8 +233,56 @@ class Repository
+    /**
+     * Create a new branch from the current branch.
+     * \param branch_name name of the new branch
+    */
```
Mis-formatted throughout the whole PR:
```suggestion
     */
```

wyrnat replied:
I hope I found all of them...

alt-graph commented on include/libgit4cpp/Repository.h:
```diff
@@ -233,8 +233,56 @@ class Repository
+     * Returns the active branch in the repository.
+     * \return shorthand name of current branch (eg. master)
+    */
+    std::string get_current_branch();
```
I am not sure about the function name here. The two `new_branch()` overloads return a `LibGitReference` to the new branch, but `get_current_branch()` returns a string? Maybe it should be `get_current_branch_name()` instead?

alt-graph commented on include/libgit4cpp/Repository.h:
```diff
@@ -233,8 +233,56 @@ class Repository
+     *  0 -> All
+     *  1 -> Local
+     *  2 -> Remote
+     * \return vector of branch longnames
```
Is there really such a thing as a "shortname" and a "longname" for a branch? If yes, how can I see them with `git` on a command line?

wyrnat replied:
I guess that command line git shows mercy by not telling us that such functions exists. Everything is handled over the short name

alt-graph replied:
Got it.

alt-graph commented on include/libgit4cpp/Repository.h:
```diff
@@ -233,8 +233,56 @@ class Repository
+     *  2 -> Remote
+     * \return vector of branch longnames
+    */
+    std::vector<std::string> list_branches(int type_flag = 0);
```
The parameter should be an enum class with a "speaking" name (e.g. `BranchType::local` or something similar).

alt-graph commented on include/libgit4cpp/Repository.h:
```diff
@@ -233,8 +233,56 @@ class Repository
+    std::vector<std::string> list_branches(int type_flag = 0);
+
+    /**
+     * Checkout a branch with options.
```
As far as I see, there are not so many options here, so maybe like this?
```suggestion
     * Check out selected files from a branch.
```

alt-graph commented on include/libgit4cpp/Repository.h:
```diff
@@ -233,8 +233,56 @@ class Repository
+     * \param branch_name The branch to checkout
+     * \param paths specifies the files to checkout
+    */
+    void checkout(const std::string& branch_name, const std::vector<std::string>& paths = {"*"});
```
I am not so sure, but doesn't a single `*` in Git expand to file/folder names in the base directory only? Wouldn't we need a double asterisk `**` to check out whole directory structures?

wyrnat replied:
On the command line, I get for "git add *" and "git add **" the same behaviour. I dont know if git overwrites the operator internally to some other functionality. But: In the tests, the single asterix works for files in the second layer directory

alt-graph replied:
Well, on the command line the shell expands the wildcards, not git. But as long as you are confident that this does the right thing, it is good enough for me. :+1:

alt-graph commented on include/libgit4cpp/wrapper_functions.h:
```diff
@@ -145,6 +145,24 @@ LibGitRepository clone(const std::string& url, const std::string& repo_path);
+ * \param branch_type Which branch type to find, enum with 1=GIT_BRANCH_LOCAL, 2=GIT_BRANCH_REMOTE, 3=GIT_BRANCH_ALL
+ * \return new git_reference object
+*/
+LibGitReference branch_create(git_repository* repo, const std::string& new_branch_name, const git_commit* starting_commit, int force);
```
Parameter `force` is not documented – should it be `branch_type`?

wyrnat replied:
I rewrote the function multiple times and forgot to adapt the docu...

alt-graph commented on include/libgit4cpp/wrapper_functions.h:
```diff
@@ -153,6 +171,43 @@ LibGitReference branch_lookup(git_repository* repo, const std::string& branch_na
 std::string branch_remote_name(git_repository* repo, const std::string& branch_name);
 
+/**
+ * Returns the human-readbale name of a reference.
```
```suggestion
 * Return the human-readable name of a reference.
```

alt-graph commented on include/libgit4cpp/wrapper_functions.h:
```diff
@@ -153,6 +171,43 @@ LibGitReference branch_lookup(git_repository* repo, const std::string& branch_na
+
+/**
+ * Iterates branches over an iterator object.
+ * \exception ref will be nullptr only if ieration object reached its end.
```
```suggestion
 * \exception ref will be nullptr only if iteration object reached its end.
```

alt-graph commented on src/Repository.cc:
```diff
@@ -554,7 +554,7 @@ void Repository::clone_repo(const std::string& url, const std::filesystem::path&
     {
         url_ = url;
-        remote_ = remote_lookup(repo.get(), "origin"); // TODO: origin korrekt?
+        remote_ = remote_lookup(repo_.get(), "origin"); // TODO: origin korrekt?
```
Wait... why? Don't we mean the repository from the `clone()` call in line 550?

wyrnat replied:
I dont know why I did this. Maybe I was looking for repo.get() to replace it with repo_.get() (I did this mistake regulary) and replaced it in the only place where repo.get() was actually correct

alt-graph commented on src/Repository.cc:
```diff
@@ -594,6 +594,98 @@ bool Repository::branch_up_to_date(const std::string& branch_name)
+        paths_as_cstr.push_back(path.c_str());
+    checkout_opts.paths.strings = const_cast<char **>(paths_as_cstr.data());
+
+    //TODO: What is the reight checkout strategy? GIT_CHECKOUT_SAFE not fitting, ..._FORCE too radical?
```
I think FORCE is OK for a start. Maybe there should be a small comment about that in the function description. Later on one could add the flags to the function itself, but I wouldn't worry about that at the moment.

alt-graph commented on src/Repository.cc:
```diff
@@ -594,6 +594,98 @@ bool Repository::branch_up_to_date(const std::string& branch_name)
+    auto last_commit = get_commit(full_branch_name);
+
+    // checkout
+    auto error = git_checkout_tree(repo_.get(), (const git_object*) last_commit.get(), &checkout_opts);
```
Is this cast necessary? If yes, it should be a C++-style cast, not a C cast.

wyrnat replied:
it is necessary. Also static_cast<> does not work for me. It gives me "invalid type conversion". Probably because it does not really know those types. C cast does not care, it converts everything.

alt-graph replied:
~That is why we should not use C-style casts – you can never be sure what they do. What you need is probably a `const_cast`.~
EDIT: Now I get it. These C people emulated inheritance by declaring `git_commit` like this:
```cpp
struct git_commit {
	git_object object;

	git_vector parent_oids;
	git_oid tree_oid;

	git_signature *author;
	git_signature *committer;

	char *message;
	char *message_short;
};
```
So in C, it is legal to cast this to a `git_object`. Sigh.

alt-graph replied:
... and in addition, they hide the declaration of `git_commit` in an implementation file so we do not have access to the `object` member. So we have no other option than using a C-style cast (or reinterpret_cast) from `git_commit*` to `git_object*`. I assume that this puts us into UB land in C++, but it will probably work nonetheless :shrug:.

alt-graph replied:
Changed it into a `reinterpret_cast` in the code and added a comment why this is necessary.

alt-graph commented on src/Repository.cc:
```diff
@@ -594,6 +594,98 @@ bool Repository::branch_up_to_date(const std::string& branch_name)
+        throw Error{ cat("switch_branch: ", git_error_last()->message) };
+    
+    // go back to original state on this branch
+    // TODO: Do we want this?
```
I think we want it this way. This is exactly what a `git checkout my_branch` on the command line does, right?

wyrnat replied:
Exactly

alt-graph commented on src/wrapper_functions.cc:
```diff
@@ -182,4 +200,62 @@ std::string branch_remote_name(git_repository* repo, const std::string& branch_n
+                                " instead of reference") };
+    return {(git_reference*) ref, git_reference_free};
+}
+*/
```
Commented-out code should not be in the final diff.

alt-graph reviewed (changes requested):
It looks great, actually. As usual I have a number of nitpicks, but almost all of them are minor, and many probably reflect my missing git knowledge.

wyrnat committed:
Lars comments

alt-graph commented on include/libgit4cpp/Repository.h:
```diff
@@ -63,6 +63,8 @@ inline std::ostream& operator<<(std::ostream& stream, const RepoState& repostate
     return stream;
 }
 
+enum class BranchType {ALL = 0, LOCAL =1, REMOTE=2};
```
According to the style guide, enumerators should be in snake_case:
```suggestion
enum class BranchType {all = 0, local =1, remote=2};
```
ALL_CAPS is reserved for macros.

alt-graph reviewed (approved):
Looks great. I have two tiny comments still open, but please fix them at your convenience and merge whenever you are ready.

alt-graph commented:
@wyrnat: I'll merge this as-is so that I can build a new version. Thanks for the PR!

alt-graph committed:
Change enumerators to snake_case

[why]
Mandated by the MCS style guide.

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
Use commented reinterpret_cast instead of C-style cast

[why]
libgit2 simulates inheritance by declaring an internal struct

struct git_commit {
	git_object object;

	git_vector parent_oids;
	git_oid tree_oid;

	git_signature *author;
	git_signature *committer;

	char *message;
	char *message_short;
};

whose declaration is not visible to library users. In C, it is legal to
cast a git_commit* to a git_object*.

[how]
We have no choice but do the same, but we spell out the reinterpret_cast
and add a comment why this mess is necessary. It is also quite likely
that this is undefined behavior for C++. Sigh.

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>
